processor/servicegraphprocessor/processor.go

```diff
@@ -288,7 +299,25 @@ func (p *serviceGraphProcessor) onExpire(e *store.Edge) {
 		zap.String("connection_type", string(e.ConnectionType)),
 		zap.Stringer("trace_id", e.TraceID),
 	)
-	stats.Record(context.Background(), statExpiredEdges.M(1))
+
+	if p.config.VirtualNodeFeatureEnabled {
+		// speculate virtual node before edge get expired.
+		if len(e.ClientService) == 0 {
+			e.ClientService = "user"
+		}
+
+		if len(e.ServerService) == 0 {
+			e.ServerService = p.getPeerHost(NeedToFindAttributes, e.Peer)
+		}
+
+		e.ConnectionType = store.VirtualNode
+
+		if e.IsComplete() {
```

Comment:
`OnComplete` is called when an edge is completed after being updated. Expired edges can't be complete, so we can remove this call.